Type reconstruction rework, part 2

slavapestov opened the pull request:
Follow-up to https://github.com/apple/swift/pull/22200 fills in more gaps in the TypeDecoder and ASTDemangler.

Progress on <rdar://problem/41115104>.

slavapestov committed:
ASTDemangler: Implement types in generic local context

If we nest a type inside a local context inside a generic type,
we have to look through the local context(s) to find the outer
generic type when stripping off generic arguments.

We don't support nominal types inside generic local context
right now, but this can happen with type aliases.

DougGregor commented on lib/Demangling/Demangler.cpp:
```diff
@@ -1458,6 +1458,20 @@ NodePointer Demangler::demangleBoundGenericType() {
+    case Node::Kind::ImplicitClosure:
+    case Node::Kind::ExplicitClosure:
+    case Node::Kind::DefaultArgumentInitializer:
+    case Node::Kind::Initializer:
```
It seems very easy for this list to go out-of-date. Should we suffer through listing out all of the cases to avoid the `default`?

slavapestov committed:
ASTDemangler: Add test for private type in extension of foreign type

This was one of the cases that TypeReconstruction did not support;
let's make sure its covered by a test.

slavapestov committed:
ASTDemangler: Add support for lowered metatypes

slavapestov committed:
ASTDemangler: Add support for lowered function types

slavapestov committed:
ASTDemangler: Fix demangling of weak/unowned/unowned(unsafe)

The check here was incorrect because we might have an optional of a
reference type as well. Just remove the check to keep things simple.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please smoke test macOS
